In [101]:
#讀入資料
import pandas as pd
Data = pd.read_csv('HW4_creditcard.csv')
Data

Time        V1        V2        V3        V4        V5        V6  \
0      143360  0.087399  1.955498 -3.009114 -0.611771  2.046245 -2.034809   
1      163005 -1.969644 -0.434381 -0.203709 -0.048957  0.395736 -1.529643   
2      151325  2.062615 -1.187470 -2.940538 -1.878575  1.946530  3.129848   
3      169962 -1.736890  1.276816  0.353397 -0.648723 -0.052977 -0.262870   
4        3760  1.401276 -0.683958 -0.915143 -1.616304  1.458234  3.274761   
...       ...       ...       ...       ...       ...       ...       ...   
28919   67659  1.208510 -1.608221  0.403127 -1.465710 -1.592717  0.066524   
28920  163646  1.997085 -0.334976 -0.305136  0.356486 -0.825710 -0.878371   
28921  151403 -0.607254  1.037853  0.180245 -0.974243  1.066338 -0.552829   
28922   58304  1.276545 -0.141506  0.278940 -0.197487 -0.382798 -0.209766   
28923   32871  1.182424 -0.457900 -1.060368 -0.331379  1.812226  3.656797   

             V7        V8        V9  ...       V21       V22       V23  \
0      1.783189 -0.513037  0.091237  ... -0.141101  0.084483 -0.294411   
1      1.029201  0.203756 -0.248148  ...  0.125004 -0.380350  0.365091   
2     -0.704875  0.659123 -0.672128  ...  0.101048  0.046789  0.135103   
3      0.223275  0.533696  0.415933  ... -0.471583 -1.238002 -0.246385   
4     -1.170329  0.710345  0.360104  ... -0.278087 -0.898279  0.081211   
...         ...       ...       ...  ...       ...       ...       ...   
28919 -1.233841  0.205880 -1.713045  ... -0.260568 -0.759445  0.073831   
28920 -0.523938 -0.037454  1.507587  ... -0.131976 -0.297754  0.369404   
28921  1.173086 -0.394738 -0.019722  ... -0.224596 -0.264097 -0.027561   
28922 -0.360697  0.080531  0.233241  ... -0.139368 -0.478909 -0.015475   
28923 -0.655162  0.883222  0.505123  ... -0.114525 -0.357949 -0.165359   

            V24       V25       V26       V27       V28  Amount  Class  
0     -0.768857 -0.046838  0.087448  0.314269  0.014855    1.82      0  
1      0.268591  0.121702 -0.128947 -0.069325 -0.061543  238.28      0  
2      0.722574  0.049329 -0.226307 -0.038544 -0.058289   89.20      0  
3      0.460213  0.454608  0.249480  0.217606  0.016917   64.94      0  
4      0.920023  0.357016 -0.508385 -0.019937  0.015447   42.00      0  
...         ...       ...       ...       ...       ...     ...    ...  
28919 -0.398662 -0.034086 -0.384216  0.020646  0.029675  132.00      0  
28920 -0.046107 -0.409260 -0.922612  0.051540 -0.029423    1.00      0  
28921 -0.463851 -0.863171  0.032642  0.070639  0.134187    5.38      0  
28922 -0.470974  0.173173  0.908495 -0.079702 -0.008598   10.00      0  
28923  1.071728  0.700666  0.553724 -0.032127  0.011588   59.94      0  

[28924 rows x 31 columns]

In [102]:
from sklearn.model_selection import train_test_split
trainData, testData, a, b = train_test_split(Data, range(Data.shape[0]), test_size=0.3, random_state=None)
print("Train Data")
trainData

Train Data


Time        V1        V2        V3        V4        V5        V6  \
16426  130617  0.643191 -3.302118 -3.674420 -0.457192 -0.313015 -0.634451   
1100   132434 -0.547505  0.798072 -0.719939 -1.129561  0.925708  0.763338   
3661    50628 -1.002922 -0.245617  1.609109 -2.508401 -1.490551 -0.615692   
4312    85309 -0.836464  1.256430  1.561911 -0.243609  0.244498 -0.390097   
21391  117524 -0.601670 -0.082572  1.165874 -0.313689  0.792720  0.218642   
...       ...       ...       ...       ...       ...       ...       ...   
27788  159255  0.165580  1.174891 -0.976623 -0.539390  1.458306 -0.998753   
12559   80092 -1.304284 -0.130240  0.116299  0.858235  0.306912  0.231862   
17549   56631  0.331924 -2.006858  0.439511 -1.019766 -1.640018 -0.094975   
21027   37029 -3.014142 -2.082325  0.112485 -2.444619  1.459454  2.026146   
14381  158712  2.076891 -1.932751  0.352858 -1.118963 -2.280691 -0.061496   

             V7        V8        V9  ...       V21       V22       V23  \
16426  1.352606 -0.569726 -1.184944  ...  1.001521  0.916711 -0.977071   
1100   0.231338  0.799204 -0.277812  ...  0.366664  1.068933 -0.101523   
3661  -0.918119  0.581482 -1.970611  ... -0.055745 -0.023227 -0.244894   
4312   0.674627  0.100781 -0.696330  ... -0.240633 -0.752051 -0.213283   
21391  0.058356  0.183997  0.696064  ...  0.276106  1.087540 -0.236523   
...         ...       ...       ...  ...       ...       ...       ...   
27788  1.647407 -0.444721 -0.621014  ...  0.259985  1.023690 -0.071832   
12559  1.473941 -0.267087 -0.329210  ... -0.038414  0.177509 -0.226820   
17549 -0.270980  0.104489  1.625883  ...  0.468514  0.626428 -0.444687   
21027  0.829261  0.241690 -1.337730  ... -0.618459 -0.821883 -0.247918   
14381 -1.868798  0.036418 -0.188598  ... -0.169992  0.187868  0.247095   

            V24       V25       V26       V27       V28  Amount  Class  
16426  0.397936  0.430087  0.189876 -0.271137  0.044006  860.00      0  
1100  -1.604148 -0.318277  0.838076  0.012324 -0.015564   11.95      0  
3661  -0.196530  0.440339 -0.130735  0.169778  0.036284   24.00      0  
4312  -0.056970  0.170607  0.136295  0.115876  0.122398    1.29      0  
21391  0.688751 -0.250957  0.646913 -0.033514 -0.030815   24.73      0  
...         ...       ...       ...       ...       ...     ...    ...  
27788  0.672965 -0.656876 -0.324260  0.491718  0.331404   14.90      0  
12559 -0.969501  0.002234 -0.226842  0.555535  0.073943  342.07      0  
17549  0.319000  0.193173  0.068059 -0.024080  0.087254  438.00      0  
21027 -2.704672  0.305059  0.644525 -0.362167  0.687891  350.79      0  
14381  0.068539 -0.447162 -0.201236  0.081398 -0.017701   76.00      0  

[20246 rows x 31 columns]

In [103]:
print("Test Data")
testData

Test Data


Time        V1        V2        V3        V4        V5        V6  \
1297    69531 -1.938737 -0.221193  2.701627  0.764642  0.434525  1.691840   
13218   78327 -0.359899  1.047426  1.168694 -0.128748  0.217463 -0.477674   
3657    76840 -0.695523  0.675099  1.936497  0.075214  0.653869 -0.733180   
2330    71331 -0.664482  1.064075  1.101463  1.480420  1.235728  0.591671   
20104   70454  1.264953 -0.094201  0.482395 -0.254505 -0.587314 -0.516439   
...       ...       ...       ...       ...       ...       ...       ...   
18806   56114  1.087529 -0.620426  0.846225  0.200569 -1.029722 -0.076061   
19185   40549 -1.902557  2.142631 -0.013088 -0.091699 -0.739316 -0.374879   
3213    36701 -0.502256  1.193438  1.276245  0.052813  0.142523 -0.996321   
20055  153414 -0.547660  0.887186  2.132011 -0.062222  0.017248  0.047353   
26135  115052  2.067078 -0.159945 -1.121287  0.416204 -0.249589 -1.215426   

             V7        V8        V9  ...       V21       V22       V23  \
1297  -0.231819  0.827839  0.177639  ... -0.119702 -0.029918  0.428254   
13218  0.575698  0.115992 -0.576627  ... -0.238505 -0.667908 -0.065440   
3657   0.893138 -0.202252 -0.381895  ...  0.064933  0.272945 -0.114686   
2330   0.733728  0.068355 -1.557627  ...  0.151469  0.404636 -0.193921   
20104 -0.331694  0.013345  0.186118  ... -0.106487 -0.335763  0.047210   
...         ...       ...       ...  ...       ...       ...       ...   
18806 -0.543697  0.055454  1.210021  ... -0.215166 -0.535819 -0.055801   
19185 -0.540593  1.496217 -0.823266  ... -0.118995 -0.518186  0.167000   
3213   0.723871 -0.153266 -0.446886  ... -0.236153 -0.622527  0.003965   
20055  0.410528  0.079346 -0.180810  ... -0.138013 -0.302187 -0.173022   
26135  0.078622 -0.244031  0.777296  ... -0.311376 -0.835187  0.351665   

            V24       V25       V26       V27       V28  Amount  Class  
1297  -0.656253 -0.161678  0.826965 -0.144409  0.059199   37.96      0  
13218 -0.063495 -0.158752  0.076343  0.239205  0.083908    2.69      0  
3657   0.385511 -0.090039 -0.636154 -0.048392 -0.091216    6.24      0  
2330  -1.012396 -0.008705  0.032418  0.005827  0.122334   18.06      0  
20104  0.060791  0.125330  0.907965 -0.076044 -0.004588    1.38      0  
...         ...       ...       ...       ...       ...     ...    ...  
18806 -0.032697  0.176789  0.964445 -0.048958  0.023846   90.77      0  
19185 -0.033939 -0.082170  0.100034  0.107577  0.006042    8.91      0  
3213   0.355863 -0.138154  0.056873  0.112382  0.105518    1.29      0  
20055  1.087293  0.210701 -0.374651  0.100729  0.092951    5.00      0  
26135 -0.066148 -0.341596  0.204960 -0.080654 -0.063984    0.99      0  

[8678 rows x 31 columns]

In [104]:
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.neural_network import MLPClassifier
from sklearn.metrics import confusion_matrix
import numpy as np
import matplotlib.pyplot as plt
#直接訓練
def fromDataframeToPrediectResult(comment, trainData, testData):
    trainLabel, testLabel = list(trainData['Class']), list(testData['Class'])
    trainData = np.array(trainData.drop(['Class'], axis = 1))
    testData = np.array(testData.drop(['Class'], axis = 1))
    #訓練
    Model = DecisionTreeClassifier(random_state=0).fit(trainData, trainLabel) if comment == "Decision" else MLPClassifier(random_state=1).fit(trainData, trainLabel)
    #預測
    ModelResult = Model.predict(testData)
    #統計結果
    print(comment, "Data with")
    
    tn, fp, fn, tp = confusion_matrix(testLabel, ModelResult).ravel()
    accuracy = (tp+tn)/(tn+fp+fn+tp)
    recall = tp/(tp+fn)
    precision = tp/(tp+fp)
    F1 = 2/(1/recall + 1/precision)
    print("accuracy:",round(accuracy*100,2),"%", end = ' ')
    print("recall:",round(recall*100,2),"%", end = ' ')
    print("precision:",round(precision*100,2),"%", end = ' ')
    print("F1-Score:",round(F1*100,2),"%")

fromDataframeToPrediectResult("Neural Network", trainData, testData)

fromDataframeToPrediectResult("Decision", trainData, testData)

Neural Network Data with
accuracy: 98.57 % recall: 0.81 % precision: 50.0 % F1-Score: 1.59 %
Decision Data with
accuracy: 99.38 % recall: 82.26 % precision: 76.12 % F1-Score: 79.07 %


In [105]:
#觀察class 1和class2的數量
print("Orignal Data Label count:")
label = list(int(i) for i in Data['Class'])
print("# Class 0:",label.count(0), "# Class 1:", label.count(1))

Orignal Data Label count:
# Class 0: 28432 # Class 1: 492


In [106]:
#改變權重
def fromDataframeToPrediectResultWithChangeArgument(comment, trainData, testData):
    trainLabel, testLabel = list(trainData['Class']), list(testData['Class'])
    trainData = np.array(trainData.drop(['Class'], axis = 1))
    testData = np.array(testData.drop(['Class'], axis = 1))
    #訓練
    Model = DecisionTreeClassifier(random_state=0,class_weight="balanced").fit(trainData, trainLabel) if comment == "Decision" else MLPClassifier(random_state=1,max_iter = 300,hidden_layer_sizes=(50,)).fit(trainData, trainLabel)
    #預測
    ModelResult = Model.predict(testData)
    #統計結果
    print(comment, "Data with")
    
    tn, fp, fn, tp = confusion_matrix(testLabel, ModelResult).ravel()
    accuracy = (tp+tn)/(tn+fp+fn+tp)
    recall = tp/(tp+fn)
    precision = tp/(tp+fp)
    F1 = 2/(1/recall + 1/precision)
    print("accuracy:",round(accuracy*100,2),"%", end = ' ')
    print("recall:",round(recall*100,2),"%", end = ' ')
    print("precision:",round(precision*100,2),"%", end = ' ')
    print("F1-Score:",round(F1*100,2),"%")

fromDataframeToPrediectResultWithChangeArgument("Neural Network", trainData, testData)

fromDataframeToPrediectResultWithChangeArgument("Decision", trainData, testData)

Neural Network Data with
accuracy: 99.23 % recall: 77.42 % precision: 71.11 % F1-Score: 74.13 %
Decision Data with
accuracy: 99.42 % recall: 76.61 % precision: 81.9 % F1-Score: 79.17 %


In [107]:
#使用SMOTE套件使資料平衡
from imblearn.over_sampling import SMOTE
resampledData, resampledLabel = SMOTE().fit_resample(Data, Data['Class'])
resampledData

Time        V1        V2        V3        V4        V5        V6  \
0      143360  0.087399  1.955498 -3.009114 -0.611771  2.046245 -2.034809   
1      163005 -1.969644 -0.434381 -0.203709 -0.048957  0.395736 -1.529643   
2      151325  2.062615 -1.187470 -2.940538 -1.878575  1.946530  3.129848   
3      169962 -1.736890  1.276816  0.353397 -0.648723 -0.052977 -0.262870   
4        3760  1.401276 -0.683958 -0.915143 -1.616304  1.458234  3.274761   
...       ...       ...       ...       ...       ...       ...       ...   
56859  129266  0.600602  2.163014 -4.685539  4.695259  0.147466 -0.773385   
56860  168559 -1.749098  1.664319 -4.462847  1.959728 -1.605880 -1.325139   
56861   52834 -1.054385 -1.541470  0.809939  0.887311 -1.245744  0.615618   
56862   58261 -0.444590  1.371407  0.560315  1.162822 -0.008386 -0.761524   
56863   85533 -2.859320  3.354828 -6.059784  3.896413 -3.367987 -0.877870   

             V7        V8        V9  ...       V21       V22       V23  \
0      1.783189 -0.513037  0.091237  ... -0.141101  0.084483 -0.294411   
1      1.029201  0.203756 -0.248148  ...  0.125004 -0.380350  0.365091   
2     -0.704875  0.659123 -0.672128  ...  0.101048  0.046789  0.135103   
3      0.223275  0.533696  0.415933  ... -0.471583 -1.238002 -0.246385   
4     -1.170329  0.710345  0.360104  ... -0.278087 -0.898279  0.081211   
...         ...       ...       ...  ...       ...       ...       ...   
56859 -1.760675  0.812576 -1.696930  ...  0.445246  0.201749  0.022059   
56860 -0.895509 -0.222507 -2.052841  ...  0.245787 -0.052998  0.477443   
56861  1.320179  0.043095 -0.170401  ...  0.610142  0.820996  1.132383   
56862  0.254420 -0.289297 -0.398576  ...  0.046379  0.511635 -0.115512   
56863 -4.020389  2.076323 -3.286173  ...  0.925726 -0.033017  0.041284   

            V24       V25       V26       V27       V28      Amount  Class  
0     -0.768857 -0.046838  0.087448  0.314269  0.014855    1.820000      0  
1      0.268591  0.121702 -0.128947 -0.069325 -0.061543  238.280000      0  
2      0.722574  0.049329 -0.226307 -0.038544 -0.058289   89.200000      0  
3      0.460213  0.454608  0.249480  0.217606  0.016917   64.940000      0  
4      0.920023  0.357016 -0.508385 -0.019937  0.015447   42.000000      0  
...         ...       ...       ...       ...       ...         ...    ...  
56859  0.219084  0.034076  0.097108  0.440933  0.187362   43.944495      1  
56860 -0.080663  0.397188  0.399915  0.345109  0.131410  376.781520      1  
56861 -0.048082 -0.264472  0.671768 -0.306210 -0.162696  503.560611      1  
56862  0.395128 -0.191682 -0.358130 -0.121285 -0.110314   16.502015      1  
56863 -0.726109  0.207966  0.038928  1.168916  0.262648  242.527305      1  

[56864 rows x 31 columns]

In [108]:
#平衡資料訓練結果
print("Resample Data Label count:")
print("# Class 0:",list(resampledLabel).count(0), "# Class 1:", list(resampledLabel).count(1), end = '\n\n')

resampledtrainData, resampledtestData, a, b = train_test_split(resampledData, range(resampledData.shape[0]), test_size=0.3, random_state=None)

fromDataframeToPrediectResult("Neural Network", resampledtrainData, resampledtestData)

fromDataframeToPrediectResult("Decision", resampledtrainData, resampledtestData)

Resample Data Label count:

# Class 0: 28432 # Class 1: 28432
Neural Network Data with
accuracy: 96.67 % recall: 96.56 % precision: 96.74 % F1-Score: 96.65 %
Decision Data with
accuracy: 99.07 % recall: 99.34 % precision: 98.79 % F1-Score: 99.07 %


# 比較訓練結果
一開始沒有經過處理的資料進行訓練可以發現accuracy 雖然很高，但是recall的表現非常的差，F1的分數只有0.159而已。

因此我決定先改變兩個模型的訓練參數:

Neural Network的class_weight改成"balanced"；Decision Tree的hide layer減少為(50,)，max_iter增加為300。

可以發現accuracy稍微升高，另外recall、precision和F1也提高不少。

最後則是透過SMOTE套件來使class 1和class 2的分布結果更加平衡，訓練的結果可以發現，accuracy、recall、precision、F1的分數都非常高。